internal/server/service.go

```diff
@@ -318,14 +322,49 @@ func (s *Service) createCertManager(hosts []string, options ServiceOptions) (Cer
 		}
 	}
 
+	hostPolicy, err := s.createAutoCertHostPolicy(hosts, options)
+	if err != nil {
+		return nil, err
+	}
+
 	return &autocert.Manager{
 		Prompt:     autocert.AcceptTOS,
 		Cache:      autocert.DirCache(options.ScopedCachePath()),
-		HostPolicy: autocert.HostWhitelist(hosts...),
+		HostPolicy: hostPolicy,
 		Client:     &acme.Client{DirectoryURL: options.ACMEDirectory},
 	}, nil
 }
 
+func (s *Service) createAutoCertHostPolicy(hosts []string, options ServiceOptions) (autocert.HostPolicy, error) {
+	if options.TLSOnDemandUrl == "" {
+		return autocert.HostWhitelist(hosts...), nil
+	}
+
+	_, err := url.ParseRequestURI(options.TLSOnDemandUrl)
+	if err != nil {
+		slog.Error("Unable to parse the tls_on_demand_url URL")
+		return nil, err
+	}
+
+	slog.Info("Will use the tls_on_demand_url URL", "url", options.TLSOnDemandUrl)
+
+	return func(ctx context.Context, host string) error {
+		slog.Info("Get a certificate", "host", host)
+
+		resp, err := http.Get(fmt.Sprintf("%s?host=%s", options.TLSOnDemandUrl, url.QueryEscape(host)))
```

Comment:
Would be worth having a timeout on this request. I don't think it needs to be user-configurable; just something reasonable like a couple seconds.

Reply:
Agreed. I'm wondering that even 2 or 2 seconds is probably too long and the TLS certificate generation will fail. 

(I'll make the modifications)

Reply:
I think a couple of seconds should work OK, it just all leads to a slower experience for the client for the first request when a new cert is first generated. But could be worth testing that out to be sure.